lib/bolt/pal.rb

```diff
@@ -45,16 +43,24 @@ def initialize(msg, details = {})
       end
     end
 
-    attr_reader :modulepath, :user_modulepath
-
     def initialize(modulepath, hiera_config, resource_types, max_compiles = Etc.nprocessors,
                    trusted_external = nil, apply_settings = {}, project = nil)
       # Nothing works without initialized this global state. Reinitializing
       # is safe and in practice only happens in tests
       self.class.load_puppet
 
-      @user_modulepath = modulepath
-      @modulepath = [BOLTLIB_PATH, *modulepath, MODULES_PATH]
+      # Build a Config::Modulepath object that can perform modulepath transformations
+      # for normal bolt and apply compilation
+      if modulepath.is_a?(Array) || modulepath.is_a?(String) || modulepath.nil?
+        @modulepath = Bolt::Config::Modulepath.new(modulepath)
+      # Specifically check the superclass, instead of the class itself,
+      # so that a child class can also be sent to PAL.
+      elsif modulepath.class.superclass == Bolt::Config::Modulepath
```

Comment:
Yeah, I could see replacing the equality with `is_a?`

As far as the awkwardness: I think we need to find a way in PAL to allow for extending the inputs so that they can match our needs in a way that we avoid things like monkeypatching or redefining PAL. I was thinking it made sense to use superclass because inheritance struck me as a reasonable implementation of the requirements of both Bolt and PE, since bolt can always use the parent, and PE can use pretty standard inheritance for its needs.

If there's another approach that you think is more appropriate I am certainly on board to look at other options though.

Reply:
IMO there isn't any extra value in having this handle explicit types vs. checking the ancestry of the class. Is there a specific concern you have with the check?

My thinking here is if we check the ancestry chain, the child classes in bolt-server/PE can do whatever they want with their names, and we don't have to make any changes in bolt if the name of the classes change. We also don't have to explicitly add checks specific to only PE in to the bolt codebase, which means in the future if another tool started building on top of bolt they could re-use this architecture.

Reply:
```rb 
@modulepath = case modulepath
              when Array, String, nil
                Bolt::Config::Modulepath.new(modulepath)
              when Bolt::Config::Modulepath
                modulepath
              else
                msg = "Type error in PAL: modulepath must be one of Array, Bolt::Config::Modulepath"
                raise Bolt::Error.new(msg, "bolt/execution-error")
              end
```

But I think it might be better to just lean into duck typing here and require the caller to provide a `Bolt::Config::Modulepath` or equivalent object and avoid explicit type checking or conversion. That might have implications for existing consumers of `Bolt::PAL` though.

Reply:
Are there any consumers other than PE and bolt?

Changing all of the bolt PAL.new calls isn't too hard